### PR TITLE
Issue #71 (code cleanup): step 3: define case classes

### PR DESCRIPTION
**Pull request by Som3**
#71

### CODE REVIEW

File: core/src/main/scala/codesearch/core/index/LanguageIndex.scala [hunk 1]
```diff
@@ -87,40 +108,39 @@ trait LanguageIndex[VTable <: DefaultTable] { self: DefaultDB[VTable] =>
...
 
-    val answer = (Process(args, None, "CSEARCHINDEX" -> indexPath.toString()) #| Seq("head", "-1001")).!!
+    Future {
+      val answer = (Process(args, None, "CSEARCHINDEX" -> indexPath.toString()) #| Seq("head", "-1001")).!!
 
-    if (answer.nonEmpty) {
-      answer.split('\n')
-    } else {
-      Array()
+      if (answer.nonEmpty) {
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: Если строка answer пустая, не вернёт ли split и так пустой массив.
То есть, весь этот `if` можно заменить на `answer.split('\n')` и вообще убрать ненужную переменную answer.

File: core/src/main/scala/codesearch/core/index/HaskellIndex.scala [hunk 2]
```diff
@@ -91,7 +68,7 @@ class HaskellIndex(val ec: ExecutionContext) extends LanguageIndex[HackageTable]
     lastVersions
   }
 
-  private def contentByURI(uri: String): Option[(String, String, Result)] = {
+  override protected def contentByURI(uri: String): Option[ContentByURI] = {
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/HaskellIndex.scala]: Тут лучше уйти от императивщины - сделать flatMap на elems.lift(1)

File: core/src/main/scala/codesearch/core/index/LanguageIndex.scala (hunk 1 above)
Comment (Som3) [reply core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: I will fix it later, thx :)

File: core/src/main/scala/codesearch/core/index/HaskellIndex.scala (hunk 2 above)
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/HaskellIndex.scala]: This isn't "ContentByURI" actually, it really just parses a csearch output line. If you're touching this code anyway, you can rename it while you're at it :slightly_smiling_face: Or leave a docstring. (And hey, even if you rename it, leave a docstring anyway.)

File: core/src/main/scala/codesearch/core/index/LanguageIndex.scala [hunk 3]
```diff
@@ -153,3 +167,13 @@ trait LanguageIndex[VTable <: DefaultTable] { self: DefaultDB[VTable] =>
   private lazy val defaultExtractor: (String, String) => Unit =
     (src: String, dst: String) => Seq("tar", "-xvf", src, "-C", dst) !!
 }
+
+object LanguageIndex {
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: docstrings please

File: core/src/main/scala/codesearch/core/index/RustIndex.scala [hunk 4]
```diff
@@ -28,25 +29,21 @@ class RustIndex(val ec: ExecutionContext) extends LanguageIndex[CratesTable] wit
...
-         .slice(math.max(page - 1, 0) * 100, page * 100)
-         .flatMap(uri => contentByURI(uri, nameToVersion))
-         .groupBy(x => (x._1, x._2))
-         .map {
-           case ((name, packageLink), results) =>
-             PackageResult(name, packageLink, results.map(_._3))
-         }
-         .toSeq
-         .sortBy(_.name))
+  override def csearch(args: SearchArguments, page: Int = 0): Future[CSearchPage] = {
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/RustIndex.scala]: You're going to merge all the implementations of `csearch` in some other PR, right?

File: web-server/app/codesearch/web/controllers/CratesSearcher.scala [hunk 5]
```diff
@@ -14,22 +15,24 @@ class CratesSearcher @Inject()(implicit val executionContext: ExecutionContext)
...
-                sources == "on",
-                page = page.toInt,
-                count,
-                callURI
-              ))
-        }
+      CratesDB.updated.flatMap(
+        updated =>
+          RustIndex()
+            .csearch(SearchArguments(query, insensitive == "on", precise == "on", sources == "on"), page.toInt)
```
Comment (neongreen) [inline web-server/app/codesearch/web/controllers/CratesSearcher.scala]: Can you use named arguments here? It's easy to mix up the order of bools.

File: web-server/app/codesearch/web/controllers/CratesSearcher.scala [hunk 6]
```diff
@@ -14,22 +15,24 @@ class CratesSearcher @Inject()(implicit val executionContext: ExecutionContext)
...
+            .map {
+              case CSearchPage(results, count) =>
+                Ok(
+                  views.html.rust_search(
+                    TimeAgo.using(updated.getTime),
```
Comment (neongreen) [inline web-server/app/codesearch/web/controllers/CratesSearcher.scala]: And here.

**Comment by neongreen** [comment]
I have nitpicked about various things that were in the code before your PR (so it's not your fault), but it'd still be nice to spend a bit of time fixing them.

File: core/src/main/scala/codesearch/core/index/RustIndex.scala (hunk 4 above)
Comment (Som3) [reply core/src/main/scala/codesearch/core/index/RustIndex.scala]: Yes, Rust has a special method `contentByURI` - after complete #74 I can merge all contentByURI into the generic method - I remove this implementation `csearch`

File: core/src/main/scala/codesearch/core/index/LanguageIndex.scala (hunk 3 above)
Comment (Som3) [reply core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: for what?

Comment (neongreen) [reply core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: For everything :trollface: 

Objects and object fields

**Pull request merged**